pyroomacoustics/libroom_src/room.cpp

```diff
@@ -276,3 +276,422 @@ int Room::image_source_shoebox(
+						
+	/* Computes the next next wall_hit position given a segment defined
+	 * by its endpoints. This method also stores the index of the wall
+	 * that contains this next hit point (we needed to pass it inside a 
+	 * Vector1i to be able to call this function from python).
+	 * 
+	 * If no wall is intersected, then the next_wall_index will be -1*/
+						
+	Eigen::VectorXf result;
+	result.resize(dim);
+	
+	next_wall_index[0] = -1;
+	
+	// Upperbound on the min distance that we could find
+	float min_dist(max_dist);
+	
+	for (size_t i(0); i < walls.size(); ++i){
+		
+		Wall w = walls[i];
+		
+		// To store the result of this iteration
+		Eigen::VectorXf temp_hit;
+		temp_hit.resize(dim);
+
+
+		// As a side effect, temp_hit gets a value (VectorXf) here
+		bool intersects = (w.intersection(start, end, temp_hit) > -1);
+		
+		//~ std::cout << "\n\n\ntemp_hit : " << temp_hit[0]<< " " << temp_hit[1]<< " " << temp_hit[2] << std::endl;
+		//~ std::cout << "start : " << start[0]<< " " << start[1]<< " " << start[2] << std::endl;
+		//~ std::cout << "end : " << end[0]<< " " << end[1]<< " " << end[2] << std::endl;
+
+		if (intersects){
+				
+			
+			
+			float temp_dist = (start-temp_hit).norm();
+			
+			if (temp_dist > libroom_eps and temp_dist < min_dist){
+				
+				min_dist = temp_dist;
+				result = temp_hit;
+				
+				// update the index of the wall that is hit
+				next_wall_index[0] = i;
+			}
+		}
+		
+	}
+	return result;
+}					
+					
+					
+bool Room::scat_ray(const Wall &last_wall,
+					const Eigen::VectorXf &hit_point,
+					float radius,
+					float scat_energy,
+					float travel_time,
+					float time_thres,
+					float sound_speed,
+					std::vector<entry> &output){
+						
+	/*
+	 * Trace a one-hop scattering ray from the last wall hit to the microphone.
+	 * In case the scattering ray can reach the microphone (no wall in
+	 * between), then we add an entry to the output passed by reference.
+	 * 
+    - last_wall: The wall object where last_hit is located
+    - radius : The radius of the circle/sphere microphone
+    - last_hit: An array of length 2 or 3 defining the last wall hit position
+    - scat_energy: The energy of the scattering ray
+    - actual_travel_time: The cumulated travel time of the ray
+    - time_thresh: The time threshold of the ray
+    - sound_speed: The speed of sound
+    * 
+    * Return : true if the scattered ray reached the mic, false otw
+	*/
+	
+	bool result = false;
+	
+	Vector1i next_wall_idx(-1);
+	
+	next_wall_hit(hit_point, mic_pos, next_wall_idx);
+	
+	
+	// No wall intersecting the direct scattered ray
+	if (next_wall_idx[0] == -1){
+		
+		VectorXf mic_hit = mic_intersection(hit_point, mic_pos, mic_pos, radius);
+		float hop_dist = (hit_point - mic_hit).norm();
+		
+		update_travel_time(travel_time, hop_dist, sound_speed);
+		
+		// Remember that the energy distance attenuation will be 
+		// performed once all the ray arrived.
+		
+		if (travel_time < time_thres){
```

Comment:
I think it might be better to stop the ray after an energy threshold rather than time. Typically we want to compute the rays up to 60 dB energy decay.

Reply:
The problem is that with the ray tracing, some parts of the RIR are composed of a lot of low-energy entries. 
So setting an energy threshold will not only shorten the tail, but it will also reduce the amplitude of some parts of the rir. Do I have to worry about this then ? Or is it juste a collateral damage ?

Reply:
Ok, you are right that we are interested in the final energy of the RIR itself, so setting the ray minimum energy to 60 dB might result in not enough energy in the tail. However, the time threshold has similar problems.

Maybe we can specify both thresholds. For rays we will indeed need to specify something smaller than 60 dB. Maybe 90 or 100 dB. The time threshold is useful if we want to limit computations.

Reply:
I am a little bit confused with this notion of dB for the energy. Do I have to set a concrete threshold and put it as a parameter (as for the time threshold ) ? Or do I need to compute the energy threshold directly from the initial energy ?

Reply:
Sorry, the decay is compared to the start energy. So when going to dB, you would compute something like
```
E_decay_db = 10 * log10(E_0 / E_hit)
```
where `E_hit` is the remaining energy of the ray hitting the receiver. Of course the computation do not need to be in dB. A value of 60 dB corresponds to attenuation of the energy by a factor `1e6`. In general, we want to stop the RIR computation when the (accumulated) energy drops below this. As you correctly pointed out, we thus need to set the threshold for a single ray lower than that. We can experiment with `1e9` or `1e10` (90 and 100 dB, resp.).